File: src/ingestion/chains/plodine.ts
/**
 * Plodine Chain Adapter
 *
 * Adapter for parsing Plodine retail chain price data files.
 * Plodine uses CSV format with semicolon delimiter and Windows-1250 encoding.
 * Store resolution is based on filename.
 *
 * Special handling: Plodine files may contain prices with missing leading zero
 * (e.g., ",69" instead of "0,69"), which this adapter handles via preprocessing.
 *
 * Plodine portal: https://www.plodine.hr/info-o-cijenama
 * URL format: https://www.plodine.hr/info-o-cijenama?date=YYYY-MM-DD
 * Download links: /cjenik/download?file=FILENAME
 */

import type { DiscoveredFile, StoreMetadata } from '../core/types'
import type { CsvColumnMapping, CsvParserOptions } from '../parsers/csv'
import { BaseCsvAdapter } from './base'
import { CHAIN_CONFIGS } from './config'

/**
 * Column mapping for Plodine CSV files (current format as of 2026).
 * Maps Plodine's column names to NormalizedRow fields.
 *
 * Sample header:
 * Naziv proizvoda;Sifra proizvoda;Marka proizvoda;Neto kolicina;Jedinica mjere;
 * Maloprodajna cijena;Cijena po JM;MPC za vrijeme posebnog oblika prodaje;
 * Najniza cijena u poslj. 30 dana;Sidrena cijena na 2.5.2025;Barkod;Kategorija proizvoda;
 */
const PLODINE_COLUMN_MAPPING: CsvColumnMapping = {
  name: 'Naziv proizvoda',
  externalId: 'Sifra proizvoda',
  brand: 'Marka proizvoda',
  unitQuantity: 'Neto kolicina',
  unit: 'Jedinica mjere',
  price: 'Maloprodajna cijena',
  unitPrice: 'Cijena po JM',
  discountPrice: 'MPC za vrijeme posebnog oblika prodaje',
  lowestPrice30d: 'Najniza cijena u poslj. 30 dana',
  // Note: anchorPrice column has dynamic date in name (e.g., "Sidrena cijena na 2.5.2025")
  // Handled via preprocessContent which normalizes the header
  anchorPrice: 'Sidrena cijena',
  barcodes: 'Barkod',
  category: 'Kategorija proizvoda',
}

/**
 * Alternative column mapping for Plodine CSV files (legacy format).
 * Some older Plodine exports may use different column names.
 */
const PLODINE_COLUMN_MAPPING_ALT: CsvColumnMapping = {
  externalId: 'Šifra',
  name: 'Naziv',
  category: 'Kategorija',
  brand: 'Marka',
  unit: 'Mjerna jedinica',
  unitQuantity: 'Količina',
  price: 'Cijena',
  discountPrice: 'Akcijska cijena',
  discountStart: 'Početak akcije',
  discountEnd: 'Kraj akcije',
  barcodes: 'Barkod',
  unitPrice: 'Cijena za jedinicu mjere',
  lowestPrice30d: 'Najniža cijena u zadnjih 30 dana',
  anchorPrice: 'Sidrena cijena',
  unitPriceBaseQuantity: 'Količina za jedinicu mjere',
  unitPriceBaseUnit: 'Jedinica mjere za cijenu',
  anchorPriceAsOf: 'Datum sidrene cijene',
}

/**
 * Plodine chain adapter implementation.
 * Extends BaseCsvAdapter with custom preprocessing for price formatting issues.
 *
 * Supports single-date discovery via setDiscoveryDate() method.
 */
export class PlodineAdapter extends BaseCsvAdapter {
  /** Date to discover files for (YYYY-MM-DD format, set by CLI before discovery) */
  private discoveryDate: string | null = null

  constructor() {
    super({
      slug: 'plodine',
      name: 'Plodine',
      supportedTypes: ['csv', 'zip'],
      chainConfig: CHAIN_CONFIGS.plodine,
      columnMapping: PLODINE_COLUMN_MAPPING,
      alternativeColumnMapping: PLODINE_COLUMN_MAPPING_ALT,
      filenamePrefixPatterns: [
        /^Plodine[_-]?/i,
        /^cjenik[_-]?/i,
        /^cjenici[_-]?/i,
      ],
      rateLimitConfig: {
        requestsPerSecond: 2,
        maxRetries: 3,
      },
    })
  }

  /**
   * Set the date to use for discovery.
   * @param date - Date in YYYY-MM-DD format
   */
  setDiscoveryDate(date: string): void {
    this.discoveryDate = date
  }

  /**
   * Discover available price files from Plodine portal.
   *
   * Plodine provides ZIP archives containing CSV files for each store.
   * URL pattern: https://www.plodine.hr/cjenici/cjenici_DD_MM_YYYY_HH_MM_SS.zip
   *
   * The portal lists all available dates. We filter by the discovery date.
   *
   * @returns Array of discovered files for the specified date
   */
  async discover(): Promise<DiscoveredFile[]> {
    const discoveredFiles: DiscoveredFile[] = []
    const seenUrls = new Set<string>()

    // Use provided date or default to today
    const targetDate = this.discoveryDate || new Date().toISOString().split('T')[0]

    // Convert YYYY-MM-DD to DD_MM_YYYY format used in Plodine filenames
    const [year, month, day] = targetDate.split('-')
    const targetDatePattern = `${day}_${month}_${year}`

    const pageUrl = this.config.baseUrl
    console.log(`[DEBUG] Fetching Plodine page: ${pageUrl}`)
    console.log(`[DEBUG] Looking for date pattern: ${targetDatePattern}`)

    try {
      const response = await fetch(pageUrl, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; PriceTracker/1.0)',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        },
      })

      if (!response.ok) {
        console.error(`Failed to fetch Plodine portal: ${response.status} ${response.statusText}`)
        console.error(`  URL: ${pageUrl}`)
        return []
      }

      const html = await response.text()

      // Extract ZIP download links: href="https://www.plodine.hr/cjenici/cjenici_DD_MM_YYYY_HH_MM_SS.zip"
      const zipPattern = /href=["'](https?:\/\/[^"']*\/cjenici\/cjenici_(\d{2}_\d{2}_\d{4})_\d{2}_\d{2}_\d{2}\.zip)["']/gi

      let match: RegExpExecArray | null
      while ((match = zipPattern.exec(html)) !== null) {
        const fileUrl = match[1]
        const fileDatePattern = match[2] // DD_MM_YYYY

        // Filter by target date
        if (fileDatePattern !== targetDatePattern) {
          continue
        }

        // Skip duplicates
        if (seenUrls.has(fileUrl)) {
          continue
        }
        seenUrls.add(fileUrl)

        // Extract filename from URL
        const filename = fileUrl.split('/').pop() || `cjenici_${fileDatePattern}.zip`

        discoveredFiles.push({
          url: fileUrl,
          filename,
          type: 'zip',
          size: null,
          lastModified: new Date(targetDate),
          metadata: {
            source: 'plodine_portal',
            discoveredAt: new Date().toISOString(),
            portalDate: targetDate,
            fileDatePattern,
          },
        })
      }

      if (discoveredFiles.length === 0) {
        console.log(`[DEBUG] No ZIP files found for date ${targetDate} (pattern: ${targetDatePattern})`)
      }
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error)
      console.error(`Error discovering Plodine files: ${errorMessage}`)
      console.error(`  URL: ${pageUrl}`)
    }

    return discoveredFiles
  }

  /**
   * Fetch a discovered file with SSL certificate handling.
   * Overrides base class to use custom agent that handles Plodine's certificate.
   */
  async fetch(file: import('../core/types').DiscoveredFile): Promise<import('../core/types').FetchedFile> {
    // Wait for rate limiting
    await this.rateLimiter.throttle()

    const response = await fetch(file.url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; PriceTracker/1.0)',
        'Accept': '*/*',
      },
    })

    if (!response.ok) {
      throw new Error(`Failed to fetch ${file.url}: ${response.status} ${response.statusText}`)
    }

    const content = await response.arrayBuffer()
    const { computeSha256 } = await import('../core/storage')
    const hash = await computeSha256(content)

    return {
      discovered: file,
      content,
      hash,
    }
  }

  /**
   * Preprocess CSV content to fix Plodine-specific formatting issues.
   * - Handles missing leading zeros in decimal values (e.g., ",69" -> "0,69")
   * - Normalizes dynamic column names (e.g., "Sidrena cijena na 2.5.2025" -> "Sidrena cijena")
   */
  protected preprocessContent(content: ArrayBuffer): ArrayBuffer {
    // Decode with Windows-1250 encoding
    const decoder = new TextDecoder(this.csvConfig.encoding)
    let text = decoder.decode(content)

    // Normalize anchor price column header (remove the dynamic date suffix)
    // "Sidrena cijena na 2.5.2025" -> "Sidrena cijena"
    text = text.replace(/Sidrena cijena na \d+\.\d+\.\d+/gi, 'Sidrena cijena')

    // Fix missing leading zeros in prices
    // Pattern matches: semicolon followed by comma and digits (;,69)
    // or start of value that's just comma and digits
    // We use semicolon as Plodine uses semicolon delimiter
    text = text.replace(/;,(\d)/g, ';0,$1')

    // Also handle case where value might be at start or in quotes
    text = text.replace(/^,(\d)/gm, '0,$1')
    text = text.replace(/",(\d)/g, '"0,$1')

    // Re-encode
    const encoder = new TextEncoder()
    return encoder.encode(text).buffer as ArrayBuffer
  }

  /**
   * Get parser options with encoding override.
   * After preprocessing, content is UTF-8.
   */
  protected getParserOptions(storeIdentifier: string): Partial<CsvParserOptions> {
    return {
      defaultStoreIdentifier: storeIdentifier,
      encoding: 'utf-8', // After preprocessing, content is UTF-8
    }
  }

  /**
   * Extract store metadata from Plodine filename.
   *
   * Example filename: SUPERMARKET_ALOJZIJA_STEPINCA_201_32100_VINKOVCI_152_246_15012026015256.csv
   * Pattern: {type}_{address...}_{postal}_{city}_{storeId}_{seq}_{date}.csv
   */
  extractStoreMetadata(file: DiscoveredFile): StoreMetadata | null {
    const parts = file.filename.replace(/\.csv$/i, '').split('_')
    if (parts.length < 6) return super.extractStoreMetadata(file)

    const storeType = parts[0]  // "SUPERMARKET"

    // Find postal code (5 digits) working from index 1
    let postalIdx = -1
    for (let i = 1; i < parts.length - 3; i++) {
      if (/^\d{5}$/.test(parts[i])) {
        postalIdx = i
        break
      }
    }

    if (postalIdx === -1) return super.extractStoreMetadata(file)

    const address = parts.slice(1, postalIdx).join(' ')
    const postalCode = parts[postalIdx]
    const city = parts[postalIdx + 1]

    return {
      name: `Plodine ${this.titleCase(city)}`,
      address: this.titleCase(address),
      city: this.titleCase(city),
      postalCode,
      storeType: this.titleCase(storeType),
    }
  }

  /**
   * Convert string to title case.
   * Example: "ALOJZIJA STEPINCA" -> "Alojzija Stepinca"
   */
  private titleCase(str: string): string {
    return str.toLowerCase().split(' ').map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(' ')
  }
}

/**
 * Create a Plodine adapter instance.
 */
export function createPlodineAdapter(): PlodineAdapter {
  return new PlodineAdapter()
}

File: services/price-service/internal/adapters/chains/plodine.go
package chains

import (
	"context"
	"fmt"
	"io"
	"regexp"
	"strings"
	"time"

	"github.com/kosarica/price-service/internal/adapters/base"
	"github.com/kosarica/price-service/internal/adapters/config"
	zipexpand "github.com/kosarica/price-service/internal/ingestion/zip"
	"github.com/kosarica/price-service/internal/parsers/csv"
	"github.com/kosarica/price-service/internal/types"
)

// plodineColumnMapping is the primary column mapping for Plodine CSV files
var plodineColumnMapping = csv.CsvColumnMapping{
	ExternalID:     types.StringPtr("Sifra proizvoda"),
	Name:           "Naziv proizvoda",
	Category:       types.StringPtr("Kategorija proizvoda"),
	Brand:          types.StringPtr("Marka proizvoda"),
	Unit:           types.StringPtr("Jedinica mjere"),
	UnitQuantity:   types.StringPtr("Neto kolicina"),
	Price:          "Maloprodajna cijena",
	DiscountPrice:  types.StringPtr("MPC za vrijeme posebnog oblika prodaje"),
	Barcodes:       types.StringPtr("Barkod"),
	UnitPrice:      types.StringPtr("Cijena po JM"),
	LowestPrice30d: types.StringPtr("Najniza cijena u poslj. 30 dana"),
	// Note: anchorPrice column has dynamic date in name, handled via preprocessing
	AnchorPrice: types.StringPtr("Sidrena cijena"),
}

// plodineColumnMappingAlt is the alternative column mapping for Plodine CSV files
var plodineColumnMappingAlt = csv.CsvColumnMapping{
	ExternalID:          types.StringPtr("Šifra"),
	Name:                "Naziv",
	Category:            types.StringPtr("Kategorija"),
	Brand:               types.StringPtr("Marka"),
	Unit:                types.StringPtr("Mjerna jedinica"),
	UnitQuantity:        types.StringPtr("Količina"),
	Price:               "Cijena",
	DiscountPrice:       types.StringPtr("Akcijska cijena"),
	DiscountStart:       types.StringPtr("Početak akcije"),
	DiscountEnd:         types.StringPtr("Kraj akcije"),
	Barcodes:            types.StringPtr("Barkod"),
	UnitPrice:           types.StringPtr("Cijena za jedinicu mjere"),
	LowestPrice30d:      types.StringPtr("Najniža cijena u zadnjih 30 dana"),
	AnchorPrice:         types.StringPtr("Sidrena cijena"),
	UnitPriceBaseQuantity: types.StringPtr("Količina za jedinicu mjere"),
	UnitPriceBaseUnit:   types.StringPtr("Jedinica mjere za cijenu"),
	AnchorPriceAsOf:     types.StringPtr("Datum sidrene cijene"),
}

// PlodineAdapter is the chain adapter for Plodine retail chain
type PlodineAdapter struct {
	*base.BaseCsvAdapter
	discoveryDate string // Date filter for discovery (YYYY-MM-DD format)
}

// NewPlodineAdapter creates a new Plodine adapter
func NewPlodineAdapter() (*PlodineAdapter, error) {
	chainConfig := config.ChainConfigs[config.ChainPlodine]

	adapterConfig := base.CsvAdapterConfig{
		BaseAdapterConfig: base.BaseAdapterConfig{
			Slug:           string(config.ChainPlodine),
			Name:           chainConfig.Name,
			SupportedTypes: []types.FileType{types.FileTypeCSV, types.FileTypeZIP},
			ChainConfig:    chainConfig,
			FilenamePrefixPatterns: []string{
				`(?i)^Plodine[_-]?`,
				`(?i)^cjenik[_-]?`,
				`(?i)^cjenici[_-]?`,
			},
			FileExtensionPattern: regexp.MustCompile(`\.(csv|CSV|zip|ZIP)$`),
		},
		ColumnMapping:            plodineColumnMapping,
		AlternativeColumnMapping: &plodineColumnMappingAlt,
	}

	baseAdapter, err := base.NewBaseCsvAdapter(adapterConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create base CSV adapter: %w", err)
	}

	return &PlodineAdapter{
		BaseCsvAdapter: baseAdapter,
	}, nil
}

// SetDiscoveryDate sets the date to use for discovery filtering
func (a *PlodineAdapter) SetDiscoveryDate(date string) {
	a.discoveryDate = date
}

// Discover discovers available Plodine price files from the portal
// Plodine provides ZIP archives containing CSV files for each store
func (a *PlodineAdapter) Discover(targetDate string) ([]types.DiscoveredFile, error) {
	discoveredFiles := make([]types.DiscoveredFile, 0)
	seenURLs := make(map[string]bool)

	// Use provided date or default to today
	date := targetDate
	if date == "" {
		date = a.discoveryDate
	}
	if date == "" {
		date = time.Now().Format("2006-01-02")
	}

	// Convert YYYY-MM-DD to DD_MM_YYYY format used in Plodine filenames
	parts := strings.Split(date, "-")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid date format: %s (expected YYYY-MM-DD)", date)
	}
	targetDatePattern := fmt.Sprintf("%s_%s_%s", parts[2], parts[1], parts[0])

	fmt.Printf("[DEBUG] Fetching Plodine page: %s\n", a.BaseURL())
	fmt.Printf("[DEBUG] Looking for date pattern: %s\n", targetDatePattern)

	resp, err := a.HTTPClient().Get(a.BaseURL())
	if err != nil {
		fmt.Printf("[ERROR] Failed to fetch Plodine portal: %v\n", err)
		return nil, fmt.Errorf("failed to fetch Plodine portal: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		fmt.Printf("[ERROR] Plodine portal returned status %d\n", resp.StatusCode)
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	html := string(bodyBytes)

	// Try multiple patterns to find ZIP files
	patterns := []*regexp.Regexp{
		// Primary: /cjenici/ path with full timestamp
		regexp.MustCompile(`href=["'(https://[^"']*\/cjenici\/cjeniki_(\d{2}_\d{2}_\d{4})_\d{2}_\d{2}_\d{2}\.zip)["']`),
		// Alternative: any path with cjenici_ prefix
		regexp.MustCompile(`href=["'](https://[^"']*\/cjenici_(\d{2}_\d{2}_\d{4})_\d{2}_\d{2}_\d{2}\.zip)["']`),
		// Relative URLs
		regexp.MustCompile(`href["']([^"']*cjenici_(\d{2}_\d{2}_\d{4})_\d{2}_\d{2}_\d{2}\.zip)["']`),
	}

	for _, zipPattern := range patterns {
		matches := zipPattern.FindAllStringSubmatch(html, -1)
		for _, match := range matches {
			if len(match) < 3 {
				continue
			}

			var fileURL, fileDatePattern string
			if strings.HasPrefix(match[1], "http") {
				fileURL = match[1]
				fileDatePattern = match[2]
			} else {
				fileURL = "https://www.plodine.hr" + match[1]
				fileDatePattern = match[2]
			}

			// Filter by target date
			if fileDatePattern != targetDatePattern {
				continue
			}

			// Skip duplicates
			if seenURLs[fileURL] {
				continue
			}
			seenURLs[fileURL] = true

			filename := fileURL[strings.LastIndex(fileURL, "/")+1:]

			lastModified, _ := time.Parse("2006-01-02", date)

			discoveredFiles = append(discoveredFiles, types.DiscoveredFile{
				URL:          fileURL,
				Filename:     filename,
				Type:         types.FileTypeZIP,
				Size:         nil,
				LastModified: types.TimePtr(lastModified),
				Metadata: map[string]string{
					"source":       "plodine_portal",
					"discoveredAt": time.Now().Format(time.RFC3339),
					"portalDate":   date,
					"fileDatePattern": fileDatePattern,
				},
			})
		}

		if len(discoveredFiles) > 0 {
			break
		}
	}

	if len(discoveredFiles) == 0 {
		fmt.Printf("[DEBUG] No ZIP files found for date %s\n", date)
	} else {
		fmt.Printf("[DEBUG] Found %d file(s) for date %s\n", len(discoveredFiles), date)
	}

	return discoveredFiles, nil
}

// ExpandZIP expands a ZIP file and returns the extracted CSV files
func (a *PlodineAdapter) ExpandZIP(ctx context.Context, content []byte, filename string) ([]zipexpand.ExpandedFile, error) {
	// Preprocess ZIP content: first expand to find CSV files, then preprocess each
	expanded, err := zipexpand.ExpandInMemory(content, filename)
	if err != nil {
		return nil, fmt.Errorf("failed to expand ZIP: %w", err)
	}

	// Preprocess each CSV file to fix price formatting
	for i := range expanded {
		if expanded[i].Type == types.FileTypeCSV {
			expanded[i].Content = a.preprocessCSVContent(expanded[i].Content)
		}
	}

	return expanded, nil
}

// preprocessCSVContent preprocesses CSV content to fix Plodine-specific formatting issues
// - Handles missing leading zeros in decimal values (e.g., ",69" -> "0,69")
// - Normalizes dynamic column names (e.g., "Sidrena cijena na 2.5.2025" -> "Sidrena cijena")
func (a *PlodineAdapter) preprocessCSVContent(content []byte) []byte {
	text := string(content)

	// Normalize anchor price column header (remove the dynamic date suffix)
	// "Sidrena cijena na 2.5.2025" -> "Sidrena cijena"
	text = regexp.MustCompile(`Sidrena cijena na \d+\.\d+\.\d+`).ReplaceAllString(text, "Sidrena cijena")

	// Fix missing leading zeros in prices
	// Pattern matches: semicolon followed by comma and digits (;,69) -> ;0,69
	text = regexp.MustCompile(`;,(\d)`).ReplaceAllString(text, ";0,$1")

	// Also handle case where value might be at start or in quotes
	text = regexp.MustCompile(`^,(\d)`).ReplaceAllString(text, "0,$1")
	text = regexp.MustCompile(`",(\d)`).ReplaceAllString(text, `"0,$1`)

	return []byte(text)
}

// Parse parses CSV content with Plodine-specific preprocessing
func (a *PlodineAdapter) Parse(content []byte, filename string, options *types.ParseOptions) (*types.ParseResult, error) {
	// Preprocess content to fix formatting issues
	preprocessed := a.preprocessCSVContent(content)

	// Use base adapter's Parse method with preprocessed content
	return a.BaseCsvAdapter.Parse(preprocessed, filename, options)
}

// ExtractStoreMetadata extracts store metadata from Plodine filename
// Pattern: {type}_{address...}_{postal}_{city}_{storeId}_{seq}_{date}.csv
func (a *PlodineAdapter) ExtractStoreMetadata(file types.DiscoveredFile) *types.StoreMetadata {
	baseName := regexp.MustCompile(`\.(csv|CSV)$`).ReplaceAllString(file.Filename, "")
	parts := strings.Split(baseName, "_")
	if len(parts) < 6 {
		storeID := a.ExtractStoreIdentifierFromFilename(file.Filename)
		return &types.StoreMetadata{
			Name: fmt.Sprintf("%s %s", a.Name(), storeID),
		}
	}

	storeType := parts[0]

	// Find postal code (5 digits) working from index 1
	postalIdx := -1
	for i := 1; i < len(parts)-3; i++ {
		if regexp.MustCompile(`^\d{5}$`).MatchString(parts[i]) {
			postalIdx = i
			break
		}
	}

	if postalIdx == -1 {
		storeID := a.ExtractStoreIdentifierFromFilename(file.Filename)
		return &types.StoreMetadata{
			Name: fmt.Sprintf("%s %s", a.Name(), storeID),
		}
	}

	address := strings.Join(parts[1:postalIdx], " ")
	postalCode := parts[postalIdx]
	city := ""
	if postalIdx+1 < len(parts) {
		city = parts[postalIdx+1]
	}

	return &types.StoreMetadata{
		Name:       fmt.Sprintf("Plodine %s", titleCase(city)),
		Address:    titleCase(address),
		City:       titleCase(city),
		PostalCode: postalCode,
		StoreType:  titleCase(storeType),
	}
}

// ExtractStoreIdentifierFromFilename extracts store identifier from Plodine filename
func (a *PlodineAdapter) ExtractStoreIdentifierFromFilename(filename string) string {
	baseName := regexp.MustCompile(`\.(csv|CSV)$`).ReplaceAllString(filename, "")
	parts := strings.Split(baseName, "_")
	if len(parts) >= 6 {
		// Store ID is typically the 5th element (index 5)
		// Pattern: {type}_{address...}_{postal}_{city}_{storeId}_{seq}_{date}
		return parts[5]
	}

	// Fallback to base class method
	dummyFile := types.DiscoveredFile{Filename: filename}
	if id := a.ExtractStoreIdentifier(dummyFile); id != nil {
		return id.Value
	}
	return ""
}

// UsesZIP returns true as Plodine uses ZIP files
func (a *PlodineAdapter) UsesZIP() bool {
	return true
}
